### PR TITLE
Add ZomeCall, InstallAppBundle (w membrane proof opt), & ActivateApp Commands

### PR DESCRIPTION
**Pull request by JettTech**
### Updates:
  - adds `installAppBundle` command - with option to provide membrane proof (:rocket:) 
  - adds `activateApp` command
  - adds `zomeCall` command

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -96,24 +98,131 @@ export async function getArgs () {
...
+        } catch (error) {
+          throw new Error('Invalid AppBundleSource. The source must be either a file or url path to the .happ file.')
+        }
+        const bundlePath = path.join(__dirname, AppBundleSource)
+        bundleSource = bundlePath
+      }
+
+      const membraneProofs = {
+        proof: Buffer.from(membraneProof, 'base64'),
+        cell_nick: cellNick // 'elemental-chat'
```
Comment (robbiecarlton) [inline src/index.js:135]:
```suggestion
        cell_nick: cellNick
```

File: src/index.js [hunk 2]
```diff
@@ -96,24 +98,131 @@ export async function getArgs () {
...
+        }
+        bundleSource = AppBundleSource
+      } else {
+        try {
+          const isHappExt = path.extname(AppBundleSource) === '.happ'
+          if (!isHappExt) throw new Error('Invalid AppBundleSource file path.  Please ensure to provide the url address of your bundle\'s .happ file.')
+        } catch (error) {
+          throw new Error('Invalid AppBundleSource. The source must be either a file or url path to the .happ file.')
+        }
+        const bundlePath = path.join(__dirname, AppBundleSource)
```
Comment (robbiecarlton) [inline src/index.js]: not sure why, but on my system, `__dirname` here is `/`, not the current directory.

File: src/index.js (hunk 1 above)
Comment (robbiecarlton) [inline src/index.js]: Latest conductor-api code looks like the format for this should be 
```suggestion
      const membraneProofs = {
        [[cell_nick]]: Buffer.from(membraneProof, 'base64'),      
```

see this line: https://github.com/holochain/holochain-conductor-api/blob/8967b93d3e3eace63110b20ea2765112e3388d54/src/api/admin.ts#L133

I haven't investigated if this is a new change in conductor-api, if it is please ignore this suggestion :)

Comment (JettTech) [reply src/index.js]: Good catch. Updated. :medal_sports:

File: src/index.js (hunk 2 above)
Comment (JettTech) [reply src/index.js]: Apparently it's a webpack thing. Updated.

File: src/index.js [hunk 3]
```diff
@@ -96,24 +98,133 @@ export async function getArgs () {
...
+        bundleSource = await downloadFile(AppBundleSource)
+      } else {
+        try {
+          const isHappExt = path.extname(AppBundleSource) === '.happ'
+          if (!isHappExt) throw new Error('Invalid AppBundleSource file path.  Please ensure to provide the url address of your bundle\'s .happ file.')
+        } catch (error) {
+          throw new Error('Invalid AppBundleSource. The source must be either a file or url path to the .happ file.')
+        }
+        // note: __dirname is configured to the 'dist' folder, not project root
+        const bundlePath = path.join(__dirname, '../', AppBundleSource)
```
Comment (zo-el) [inline src/index.js:136]: What is the reasoning behind this?

**Review by zo-el — COMMENTED** [review]
Three things:
- Can we have the code structured more modular? Just pulling the command functions into their own folders would make it more convenient.
- Also we would need to have the ability to set properties too. But we can just add it later.
- Question about the path. Can we just expect an absolute path?

File: src/index.js [hunk 4]
```diff
@@ -96,24 +98,133 @@ export async function getArgs () {
...
+  program
+    .command('installAppBundle <InstalledAppId> <AgentHash> <AppBundleSource>')
+    .alias('b')
+    .option('-e --membraneProof <membraneProof>', 'provide membrane proof for app bundle install and runtime validation - should be paired with --cellNick')
+    .option('-n --cellNick <cellNick>', 'provide cell nick - should be paired with --membraneProof')
+    .option('-u --uid <uid>', 'provide uid to happ')
+    .description('install provided happ bundle with given id and details: calls InstallAppBundle(installed_app_id, agent_key, source, membrane_proofs?, uid?) -> { installed_app_id, cell_data: [ { cell_id, cell_nick } ], status: { inactive: { reason: [Object] } } }')
+    .action(async (InstalledAppId, AgentHash, AppBundleSource, { membraneProof, cellNick, uid }) => {
+      let membraneProofs
+      // throw error if one, but now both, of the membraneProof and CellNick are present - both are needed to form the membrane_proofs object
```
Comment (robbiecarlton) [inline src/index.js]:
```suggestion
      // throw error if one, but not both, of the membraneProof and CellNick are present - both are needed to form the membrane_proofs object
```

File: src/index.js [hunk 5]
```diff
@@ -96,24 +98,133 @@ export async function getArgs () {
...
+      if (!membraneProof ^ !cellNick) {
```
Comment (robbiecarlton) [inline src/index.js:111]:
```suggestion
      if (!(membraneProof || cellNick)) {
```

Feel free to ignore this suggestion, but I find it easier to read written this way. It reads as an english sentence "if not membraneProof or cellNick" whereas using the exclusive or version, there isn't a simple translation into english so I have to do more thinking to read it.

File: src/index.js [hunk 6]
```diff
@@ -96,24 +98,133 @@ export async function getArgs () {
...
+    .description('call zome function for cell: calls callZome(cell_id, zome_name, fn_name, payload, provenence) -> ZomeCallResult: any')
+    .action(async (DnaHash, AgentHash, ZomeName, ZomeFunction, Payload) => {
+      let payload = null
+      function toJSONString (input) {
+        const keyMatcher = '([^",{}\\s]+?)'
+        const valMatcher = '(?<=: )(.*?)(?=,|}| })'
+        const matcher = new RegExp(`${keyMatcher}\\s*:\\s*${valMatcher}`, 'g')
+        const parser = (_, key, value) => {
+          console.log('key ', key)
+          console.log('value ', value)
```
Comment (robbiecarlton) [inline src/index.js:187]: do we want to keep these logs?

**Review by zo-el — APPROVED** [review]

File: src/utils.js [hunk 7]
```diff
@@ -41,6 +45,84 @@ const stringifyBuffRec = (obj) => {
   return obj
 }
 
+/**
+ * Downloads url and saves to tmp file
+ * @returns {filePath}
+ */
+export const downloadFile = async (downloadUrl) => {
+  console.log('Downloading url: ', downloadUrl)
```
Comment (robbiecarlton) [inline src/utils.js:53]: this log is duplicated at the call site. Consider removing it here and just leaving it there.

File: src/utils.js [hunk 8]
```diff
@@ -41,6 +45,84 @@ const stringifyBuffRec = (obj) => {
...
+  urlObj.protocol = 'https'
+  downloadUrl = urlObj.toString()
+
+  return new Promise((resolve, reject) => {
+    request({
+      uri: downloadUrl
+    })
+      .pipe(file)
+      .on('finish', () => {
+        // console.log(`Downloaded file from ${downloadUrl} to ${fileName}`);
```
Comment (robbiecarlton) [inline src/utils.js:68]: remove comment

File: src/index.js (hunk 3 above)
Comment (JettTech) [reply src/index.js:136]: That is bc in the webpack config, the root source is set to the dist folder, so in order to get the path from the project directory for the app path, we will always need to go out one.

**Review by robbiecarlton — APPROVED** [review]
A few suggestions and stray logs and comments. Once you've looked at them and decided if you want to take them or not, I don't need to review again.

File: src/index.js (hunk 6 above)
Comment (JettTech) [reply src/index.js:187]: These can be removed.

**Comment by JettTech** [comment]
@zo-el 
> Can we have the code structured more modular? Just pulling the command functions into their own folders would make it more convenient.
> Also we would need to have the ability to set properties too. But we can just add it later.

This is a fine request, but would like to do these together as a separate PR

**Pull request merged**